pkg/cnsmigration/migrator.go

```diff
@@ -0,0 +1,315 @@
+		printGreenInfo("Starting migration for pv %s", pvName)
+		pv, err := c.clientSet.CoreV1().PersistentVolumes().Get(ctx, pvName, metav1.GetOptions{})
+		if err != nil {
+			msg := fmt.Errorf("error finding pv %s: %v", pvName, err)
+			printErrorObject(msg)
+			return msg
+		}
+
+		csiSource := pv.Spec.CSI
+		if csiSource != nil && csiSource.Driver == vSphereCSIDriverName {
+			if c.checkForDatastore(pv) {
+				err = c.migrateVolume(ctx, csiSource.VolumeHandle)
+				if err != nil {
+					c.failedToMigrate++
+					printError("error migrating volume %s with error: %v", pvName, err)
+				} else {
+					c.migratedVolumes++
+					printGreenInfo("successfully migrated pv %s with volumeID: %s", pvName, csiSource.VolumeHandle)
+				}
+			}
+		} else {
+			printGreenInfo("pv %s is not a vSphere CSI volume", pvName)
+		}
+	}
+	return nil
+}
+
+func (c *CNSVolumeMigrator) migrateVolume(ctx context.Context, volumeID string) error {
+	dtsDatastore, err := c.getDatastore(ctx, c.destinationDatastore)
+	if err != nil {
+		msg := fmt.Errorf("error finding destinationd datastore %s: %v", dtsDatastore, err)
+		printErrorObject(msg)
+		return msg
+	}
+
+	relocatedSpec := types.NewCnsBlockVolumeRelocateSpec(volumeID, dtsDatastore.Reference())
+	task, err := c.vSphereConnection.CnsClient().RelocateVolume(ctx, relocatedSpec)
+	if err != nil {
+		// Handle case when target DS is same as source DS, i.e. volume has
+		// already relocated.
+		if soap.IsSoapFault(err) {
+			soapFault := soap.ToSoapFault(err)
+			printError("type of fault: %v. SoapFault Info: %v", reflect.TypeOf(soapFault.VimFault()), soapFault)
+			_, isAlreadyExistErr := soapFault.VimFault().(vim.AlreadyExists)
+			if isAlreadyExistErr {
+				// Volume already exists in the target SP, hence return success.
+				return nil
+			}
+		}
+		return err
+	}
+	taskInfo, err := task.WaitForResultEx(ctx)
+	if err != nil {
+		msg := fmt.Errorf("error waiting for relocation task: %v", err)
+		printErrorObject(msg)
+		return msg
+	}
+	results := taskInfo.Result.(types.CnsVolumeOperationBatchResult)
+	for _, result := range results.VolumeResults {
+		fault := result.GetCnsVolumeOperationResult().Fault
+		if fault != nil {
+			printError("Fault: %+v encountered while relocating volume %v", fault, volumeID)
+			return fmt.Errorf(fault.LocalizedMessage)
+		}
+	}
+	return nil
+}
+
+func (c *CNSVolumeMigrator) checkForDatastore(pv *v1.PersistentVolume) bool {
+	csiSource := pv.Spec.CSI
+	for _, cnsVolume := range c.matchingCnsVolumes {
+		vh := csiSource.VolumeHandle
+		if cnsVolume.VolumeId.Id == vh {
+			printGreenInfo("found a volume to migrate: %s", vh)
+			pvcName, podName, inUseFlag := c.usedVolumeCache.volumeInUse(volumeHandle(vh))
+			if inUseFlag {
+				c.failedToMigrate++
+				printError("volume %s is being used by pod %s in pvc %s", vh, podName, pvcName)
+				return false
+			}
+			return true
+		}
+	}
+	c.volumesNotFound++
+	printInfo("Unable to find volume %s in CNS datastore %s", pv.Name, c.sourceDatastore)
```

Comment:
Printing `volumeHandle` or something to look for in the datastore would be helpful.